### PR TITLE
feat(pgwire):support portal description 

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Support portal type in describe message. It is required when using JDBC.

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Add the 'user-facing changes' label if your PR contains changes that are visible to users (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#3392

### CODE REVIEW

File: src/utils/pgwire/src/pg_extended.rs [hunk 1]
```diff
@@ -123,16 +123,33 @@ impl PgStatement {
...
+        session: Arc<SM::Session>,
+        portal_name: String,
+        params: &[Bytes],
+    ) -> Result<PgPortal, ()> {
         let statement = cstr_to_str(&self.query_string).unwrap().to_owned();
 
         if params.is_empty() {
-            return PgPortal {
+            let row_description =
+                if statement.starts_with("select") || statement.starts_with("SELECT") {
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_extended.rs:136]: I' m not sure whether query can be `sElEcT`

File: src/utils/pgwire/src/pg_extended.rs [hunk 2]
```diff
@@ -123,16 +123,33 @@ impl PgStatement {
...
+                    if let Ok(rows) = session.infer_return_type(statement.as_str()).await {
+                        rows
+                    } else {
+                        return Err(());
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_extended.rs:140]: What about use `?`

File: src/utils/pgwire/src/pg_extended.rs [hunk 3]
```diff
@@ -155,13 +172,29 @@ impl PgStatement {
         // "SELECT 'A', 'B', 'C'".
         let instance_query_string = replace_params(statement, &generic_params, &params);
 
-        // 4. Create a new portal.
-        PgPortal {
+        // 4. Get row_description and return portal.
+        let row_description = if instance_query_string.starts_with("select")
+            || instance_query_string.starts_with("SELECT")
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_extended.rs:177]: This part of code is very similar to above.

File: src/utils/pgwire/src/pg_protocol.rs [hunk 4]
```diff
@@ -273,21 +282,37 @@ where
...
+                    } else {
+                        // NOTE Error handle need modify later.
+                        named_portals.get(&name).unwrap()
+                    };
+
+                    self.write_message(&BeMessage::RowDescription(&portal.row_desc()))
+                        .await?;
+                } else {
+                    // TODO: Error Handle
+                    todo!()
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_protocol.rs:314]: Link a TODO issue in code. You can check https://github.com/singularity-data/risingwave/issues/2366. Investigate whether we can improve Error in the pgwire-crates

File: src/utils/pgwire/src/pg_extended.rs (hunk 2 above)
Comment (ZENOTME) [reply src/utils/pgwire/src/pg_extended.rs:140]: > What about use `?`

infer_return_type -> Result<-, BoxedError>
We can't covert Result<-, BoxedError> to Result<-,()> when using '?'. Or is there another way can process it more concisely?

File: src/utils/pgwire/src/pg_protocol.rs [hunk 5]
```diff
@@ -245,7 +245,11 @@ where
 
                 // 2. Instance the statement to get the portal.
                 let portal_name = cstr_to_str(&m.portal_name).unwrap().to_string();
-                let portal = statement.instance(portal_name.clone(), &m.params);
+                let session = self.session.clone().unwrap();
+                let portal = statement
+                    .instance::<SM>(session, portal_name.clone(), &m.params)
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_protocol.rs:250]: directly self.session.clone(). No need for a temporal variable here.

File: src/utils/pgwire/src/pg_protocol.rs [hunk 6]
```diff
@@ -273,21 +282,37 @@ where
...
             FeMessage::Describe(m) => {
                 // 1. Get statement.
-                let name = cstr_to_str(&m.query_name).unwrap().to_string();
-                let statement = if name.is_empty() {
-                    unnamed_statement
-                } else {
-                    // NOTE Error handle need modify later.
-                    named_statements.get(&name).unwrap()
-                };
+                if m.kind == b'S' {
```
Comment (BowenXiao1999) [inline src/utils/pgwire/src/pg_protocol.rs:285]: Add some comments explain what is 'S' or 'P' ?

File: src/utils/pgwire/src/pg_extended.rs (hunk 2 above)
Comment (BowenXiao1999) [reply src/utils/pgwire/src/pg_extended.rs:140]: try .map_err?

File: src/utils/pgwire/src/pg_extended.rs (hunk 1 above)
Comment (neverchanje) [inline src/utils/pgwire/src/pg_extended.rs:136]: I guess you should take the first 6 characters and case-insensitively compare them with "select", which means you can convert all characters to lower-case. I don't know if "Select" would work?

File: src/utils/pgwire/src/pg_extended.rs (hunk 3 above)
Comment (ZENOTME) [reply src/utils/pgwire/src/pg_extended.rs:177]: We can extract a function inside portal?
`fn infer_row_description(sql: &str,session: Arc<SM::Session>,) -> Result<Vec<PgFieldDescriptor>, ()>`

File: src/utils/pgwire/src/pg_extended.rs [hunk 7]
```diff
@@ -123,16 +123,36 @@ impl PgStatement {
...
-    pub fn instance(&self, portal_name: String, params: &[Bytes]) -> PgPortal {
+    async fn infer_row_description<SM: SessionManager>(
+        session: Arc<SM::Session>,
+        sql: &str,
+    ) -> Result<Vec<PgFieldDescriptor>, ()> {
+        if sql.len() > 6 && sql[0..6].eq_ignore_ascii_case("select") {
+            session.infer_return_type(sql).await.map_err(|_e| ())
+        } else {
+            Ok(vec![])
+        }
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply src/utils/pgwire/src/pg_extended.rs:134]: Would it be more efficient

[the remark this answers is not in the document]
Comment (BowenXiao1999) [reply src/utils/pgwire/src/pg_extended.rs:134]: More clean. We should redcue the if-else pairs. One if is more readable than if-else

**Pull request merged**